### DISCUSSION TITLE
DragDrop cancel event

### DISCUSSION THREAD

**Discussion by jonchardy** [AvaloniaUI/Avalonia#8341] [Q&A]
Is there any event one can listen for to determine if a drag and drop operation is cancelled via an 'ESC' key input? I've added handlers for other DragDrop events, but there doesn't seem to be anything analogous to the [QueryDragContinueEvent](https://docs.microsoft.com/en-us/dotnet/api/system.windows.dragdrop.querycontinuedrag?view=windowsdesktop-6.0).

```cs
AddHandler(DragDrop.DropEvent, ...);
AddHandler(DragDrop.DragOverEvent, ...);
AddHandler(DragDrop.DragEnterEvent, ..);
AddHandler(DragDrop.DragLeaveEvent, ...);
```

During a drag and drop, KeyDown events are swallowed, so that can't be used to perform additional cancellation logic.

**Comment by cyraid** [comment]
I know this has been a long time, but anyone coming here with the new Avalonia 11, you use `DragDrop.DoDragDropAsync`. I've done:

```csharp
  private async void OnViewPointerMoved(object? sender, PointerEventArgs ev) {

    if (startDragPos is null)
      return;

    var pos = ev.GetPosition(this);

    if (Point.Distance(pos, startDragPos.Value) < 5)
      return;

    var dragResult = await DragDrop.DoDragDropAsync(ev, new DataTransfer(), DragDropEffects.Move);

    if (dragResult == DragDropEffects.None)
      DraggingCancelled();

    StopDragging();

  } // Procedure //
```

Of course in my scenario, the `TreeView` was eating the events I needed to put in the constructor:

```csharp
    AddHandler(PointerPressedEvent, OnViewPointerPressed, RoutingStrategies.Tunnel, true);
    AddHandler(PointerMovedEvent, OnViewPointerMoved, RoutingStrategies.Tunnel, true);
    AddHandler(PointerReleasedEvent, OnViewPointerReleased, RoutingStrategies.Tunnel, true);
```

`dragResult` returns DragDropEffects.None if cancelled, and your `DragEffects` you set in the OnDragOver control event (in my case `DragDropEffects.Move`).
